tools/tpm2_policycountertimer.c

```diff
@@ -0,0 +1,198 @@
+            ctx.operation = TPM2_EO_UNSIGNED_LT;
+        } else if (!strcmp(argv[0], "sge")) {
+            ctx.operation = TPM2_EO_SIGNED_GE;
+        } else if (!strcmp(argv[0], "uge")) {
+            ctx.operation = TPM2_EO_UNSIGNED_GE;
+        } else if (!strcmp(argv[0], "sle")) {
+            ctx.operation = TPM2_EO_SIGNED_LE;
+        } else if (!strcmp(argv[0], "ule")) {
+            ctx.operation = TPM2_EO_UNSIGNED_LE;
+        } else if (!strcmp(argv[0], "bs")) {
+            ctx.operation = TPM2_EO_BITSET;
+        } else if (!strcmp(argv[0], "bc")) {
+            ctx.operation = TPM2_EO_BITCLEAR;
+        } else {
+            LOG_ERR("Invalid Comparison-Operation specified.");
+            return false;
+        }
+
+    return true;
+}
+
+static bool process_input_operand_b(char *value) {
+
+    char *input_file = strcmp("-", value) ? value : NULL;
+    bool result;
+
+    if (input_file) {
+        result = files_get_file_size_path(value,
+                (long unsigned *) &ctx.operand_b.size);
+    }
+    if (input_file && !result) {
+        return false;
+    }
+
+    result = files_load_bytes_from_buffer_or_file_or_stdin(NULL, input_file,
+            (UINT16 *)&ctx.operand_b.size,
+            (unsigned char *)&ctx.operand_b.buffer);
+    if (!result) {
+        return false;
+    }
+
+    return true;
+}
+
+static bool process_input_offset(char *value) {
+
+    bool result = tpm2_util_string_to_uint16(value, &ctx.offset);
+    if (!result) {
+        LOG_ERR("Could not convert starting offset, got: \"%s\"", value);
+    }
+
+    return result;
+}
+
+static bool on_option(char key, char *value) {
+
+    bool result = true;
+
+    switch (key) {
+    case 'L':
+        ctx.policy_digest_path = value;
+        break;
+    case 'S':
+        ctx.session_path = value;
+        break;
+    case 'i':
+        result = process_input_operand_b(value);
+        goto on_option_out;
+    case 0:
+        result = process_input_offset(value);
+        goto on_option_out;
+    default:
+        return false;
+    }
+
+on_option_out:
+    return result;
+}
+
+bool tpm2_tool_onstart(tpm2_options **opts) {
+
+    const struct option topts[] = {
+        { "policy",    required_argument, NULL, 'L' },
+        { "session",   required_argument, NULL, 'S' },
+        { "input",     required_argument, NULL, 'i' },
+        { "offset",    required_argument, NULL,  0  },
+    };
+
+    *opts = tpm2_options_new("L:S:i:", ARRAY_LEN(topts), topts, on_option,
+            on_arg, 0);
+
+    return *opts != NULL;
+}
+
+bool is_input_option_args_valid(void) {
+
+    if (!ctx.session_path) {
+        LOG_ERR("Must specify -S session file.");
+        return false;
+    }
+
+    if (!ctx.operand_b.size) {
+        LOG_ERR("Data to compare is of size 0");
+        return false;
+    }
+
+    /*
+     * Ensure input data at offset is smaller than the TPMS_TIME_INFO size.
+     */
+    if ((ctx.operand_b.size + ctx.offset) > sizeof(TPMS_TIME_INFO)){
```

Comment:
why TPMS_TIME_INFO, is this something in the spec? Can you quote the source?

Reply:
The comparison in this command (TPM2_PolicyCounterTimer) is in the TPMS_TIME_INFO structured data.

Reply:
Ahh looking at the tests it is more clear now, you create some string in milliseconds and pass through XXD converting it to binary and then copy it into the structure. One, your check is wrong, you want to check against the size of the time portion of the field, which is a U64 right? I don't see the utility in setting the TPMS_TIME_INFO structure like this and requiring an offset.

I would just make time in miliseconds an argument to the tool and make the comparison operator a required option.

Then in the future, if people need to adjust other portions of the TPMs_TIME_INFO, we can set them by name:

```bash
tpm2_policycountertimer --gte 123456
tpm2_policycountertimer --gte time=123456
tpm2_policycountertimer --gte clock=123456 reset_count=2, etc.
```

I think for right now, I would just do milliseconds as an argument and populate time only.

Reply:
> Ahh looking at the tests it is more clear now, you create some string in milliseconds and pass through XXD converting it to binary and then copy it into the structure. One, your check is wrong, you want to check against the size of the time portion of the field, which is a U64 right? I don't see the utility in setting the TPMS_TIME_INFO structure like this and requiring an offset.
> 
> I would just make time in miliseconds an argument to the tool and make the comparison operator a required option.
> 
> Then in the future, if people need to adjust other portions of the TPMs_TIME_INFO, we can set them by name:
> 
> ```shell
> tpm2_policycountertimer --gte 123456
> tpm2_policycountertimer --gte time=123456
> tpm2_policycountertimer --gte clock=123456 reset_count=2, etc.
> ```
> 
> I think for right now, I would just do milliseconds as an argument and populate time only.

Although you can only specify one operand at a time as a single input operand buffer and an offset is ultimately landing in the esapi call

Reply:
> > I think for right now, I would just do milliseconds as an argument and populate time only.
> 
> Although you can only specify one operand at a time as a single input operand buffer and an offset is ultimately landing in the esapi call

Yeah but we're not trying to expose ESAPI directly, we're trying to create tools that mortals can use

Reply:
To clarify if multiple of these conditions have to be true, then it can be combined in an OR or AND fashion by calling the tool to extend the same session context, just like other policy tools.